More reliable way to measure wait time.

simon-shopify opened the pull request:
This moves the benchmarking code for ticket acquisition as close as possible to the system call to eliminate anything that could artificially inflate the numbers.

csfrancis commented on ext/semian/sysv_semaphores.h:
```diff
@@ -84,7 +84,13 @@ create_semaphore(int key, long permissions, int *created);
+  short op,
+  short flags,
+  struct timespec *ts,
+  int *elapsed_ms);
```
Should be `wait_time_ms`, I think.

csfrancis commented on ext/semian/sysv_semaphores.c:
```diff
@@ -80,19 +81,27 @@ set_semaphore_permissions(int sem_id, long permissions)
   }
+  if (wait_time_ms) {
+    clock_gettime(CLOCK_MONOTONIC, &end);
+    *wait_time_ms = begin.tv_sec * 1e3 + begin.tv_nsec / 1e6 - end.tv_sec * 1e3 - end.tv_nsec / 1e6;
```
I find this harder to understand without brackets.

dalehamel replied:
This line isn't very descriptive, I'd recommend refactoring to a static function to better describe the nature of this (relatively complicated) timing operation.

Alternatively, if you're concerned about speed, an inline could make sense - but i'd been avoiding them where possible.

csfrancis commented on ext/semian/resource.c:
```diff
@@ -63,7 +64,10 @@ semian_resource_acquire(int argc, VALUE *argv, VALUE self)
 
-  return rb_ensure(rb_yield, self, cleanup_semian_resource_acquire, self);
+  VALUE options = rb_hash_new();
+  rb_hash_aset(options, ID2SYM(rb_intern("wait_time")), INT2NUM(wait_time_ms));
```
It would be more #webscale to cache the `rb_intern` in a global.

dalehamel commented on ext/semian/sysv_semaphores.c:
```diff
@@ -80,19 +81,27 @@ set_semaphore_permissions(int sem_id, long permissions)
   buf.sem_flg = flags;
 
+  struct timespec begin, end;
+  if (wait_time_ms) {
```
I'd suggest moving both this and the timing finish to a helper function, as this complicates the control flow for a bit

dalehamel commented on ext/semian/resource.c:
```diff
@@ -54,7 +54,8 @@ semian_resource_acquire(int argc, VALUE *argv, VALUE self)
 
   /* release the GVL to acquire the semaphore */
-  acquire_semaphore_without_gvl(&res);
+  int wait_time_ms = 0;
```
please move declaration to the top

dalehamel replied:
actually, why don't we just add this as a member to the struct?

dalehamel commented on ext/semian/sysv_semaphores.c:
```diff
@@ -80,19 +81,27 @@ set_semaphore_permissions(int sem_id, long permissions)
 
 int
-perform_semop(int sem_id, short index, short op, short flags, struct timespec *ts)
+perform_semop(int sem_id, short index, short op, short flags, struct timespec *ts, int *wait_time_ms)
```
I think it could make sense to just define a type which includes both timespec and an additional int for wait_time_ms, though either way the call signature would have to change.

I think having a single argument for timing related info is cleaner, but I also thing the separate variable might be more apparent. 

something to watch though, as this function alright takes a lot of parameters.

dalehamel commented on ext/semian/sysv_semaphores.c:
```diff
@@ -130,22 +139,29 @@ get_semaphore(int key)
+
 void *
-acquire_semaphore_without_gvl(void *p)
+acquire_semaphore_without_gvl(semian_resource_t *res, int *wait_time_ms)
```
Is the main purpose of this to figure out how long we are stuck waiting on a syscall without the GVL?

dalehamel commented on ext/semian/sysv_semaphores.c:
```diff
@@ -130,22 +139,29 @@ get_semaphore(int key)
+acquire_semaphore_without_gvl(semian_resource_t *res, int *wait_time_ms)
 {
-  WITHOUT_GVL(acquire_semaphore, p, RUBY_UBF_IO, NULL);
+  struct acquire_semaphore_args args = { .res = res, .wait_time_ms = wait_time_ms };
```
might be cleaner to just include this as a member of semian_resource_t - and we actually own that type, so we don't have to do any sort of struct-wrapping.

dalehamel commented on ext/semian/sysv_semaphores.c:
```diff
@@ -130,22 +139,29 @@ get_semaphore(int key)
 {
-  semian_resource_t *res = (semian_resource_t *) p;
+  struct acquire_semaphore_args *args = p;
+  semian_resource_t *res = args->res;
```
why is this declaration needed?

dalehamel commented on ext/semian/sysv_semaphores.h:
```diff
@@ -84,7 +84,13 @@ create_semaphore(int key, long permissions, int *created);
 // The call may be timed or untimed
 int
-perform_semop(int sem_id, short index, short op, short flags, struct timespec *ts);
+perform_semop(
```
Yeah, another sign the call sig is getting complicated

dalehamel commented on ext/semian/sysv_semaphores.h:
```diff
@@ -104,7 +110,7 @@ get_semaphore(int key);
 // Decrements the ticket semaphore within the semaphore set
 void *
-acquire_semaphore_without_gvl(void *p);
+acquire_semaphore_without_gvl(semian_resource_t *p, int *wait_time_ms);
```
Are we sure it will always be a semian_resource_t? (haven't traced the code).

simon-shopify commented:
> Try and come up with a test case where you could prove the timing was inaccurate

I'll try to explain a bit more what I am doing. So I suspect the timing is accurate. However, the numbers are far off. Given that the original PR was a quick one and even I thought it was a bit sloppy, I am taking the time to do it right on the off chance that something like

* the GC,
* a rogue statsd event fired,
* some other syscall like [the cleanup semop](https://github.com/Shopify/semian/blob/4606b13ff08bd542e0fef6e19ca61d48f44e5c90/ext/semian/resource.c#L242)

is causing this delay and I can save myself a few days of work. It is really about improving the code. If you feel that this PR (once comments are addressed) is making Semian worse, I don't have a proof that it makes the measurements more accurate, so I'll just close it.

alexsnaps commented:
Random thought, could the GVL be contended when we reacquire it (e.g. GC)?

csfrancis commented:
> Random thought, could the GVL be contended when we reacquire it (e.g. GC)?

That seems unlikely for Unicorn workers since they're effectively single threaded (there are some background threads that do things like New Relic). But perhaps GC.

dalehamel commented:
> I'll try to explain a bit more what I am doing. So I suspect the timing is accurate. However, the numbers are far off.

Can you clarify what you mean? For me (without context) it sounds like you're contradicting yourself.

> Given that the original PR was a quick one and even I thought it was a bit sloppy

I think it's great to get scrappy WIP PRs up, I understood what you are trying to do and just thought I would offer tips on how to integrate this cleanly - I hope I didn't deter you.

simon-shopify commented:
> Can you clarify what you mean? For me (without context) it sounds like you're contradicting yourself.

What I meant is I think the numbers reflect reality, but these numbers are far off what I expected. I need to investigate

* if this discrepancy is real;
* if so, what is the cause.

I need to eliminate the uncertainty of the former before doing the latter, otherwise I might waste my time profiling something that doesn't exist.

> I think it's great to get scrappy WIP PRs up, I understood what you are trying to do and just thought I would offer tips on how to integrate this cleanly - I hope I didn't deter you.

I appreciate the tips, I just want to clear the issue with proving the timing are off before spending more time on this.

simon-shopify commented:
Ok, I updated the PR. I extracted the concept of **operation** from the concept of **resource**. It contains information about a single Semian operation like whether it errored and performance measurements.

Can I get another round of reviews? I'd appreciate if you kept the feedback as a list of action items, and make clear what is a blocker or not to merge this PR.

csfrancis commented on ext/semian/resource.c:
```diff
@@ -239,9 +256,7 @@ cleanup_semian_resource_acquire(VALUE self)
   semian_resource_t *res = NULL;
   TypedData_Get_Struct(self, semian_resource_t, &semian_resource_type, res);
-  if (perform_semop(res->sem_id, SI_SEM_TICKETS, 1, SEM_UNDO, NULL) == -1) {
-    res->error = errno;
```
Why did you get rid of this?

simon-shopify commented:
> I feel that this implementation is adding too much code for what it needs to do. Do we really need to support multiple benchmark topics when we're only currently just using one?
> 
> Why not just add a wait_time member to semian_resource_t?

Sure, I can change it back. What do you think, @dalehamel ?

dalehamel commented on ext/semian/sysv_semaphores.c:
```diff
@@ -130,24 +131,28 @@ get_semaphore(int key)
 }
 
 static void *
 acquire_semaphore(void *p)
```
I assume this was originally a semian_resource_t? If so, then this should be a fair bit simpler if you just add the timing as a member to that struct.

dalehamel commented:
> Why not just add a wait_time member to semian_resource_t?

Yeah I think I suggested something along these lines originally. That's what I was trying to get at with:

> Can some of the arguments that have been added instead just become struct members, where a struct is already being passed especially if we own the struct type? For cases where we don't (timespec) should we wrap it with a custom type?

The code here is well written and well organized, but I think Scott's critique is on-point that just adding another member to this struct is probably a simpler way.

I think you can refactor your timing helper functions to just be static members of whatever file you are doing the timing in.

My apologies for not being able to give a very detailed review here. Please just take my comments as suggestions and not requirements.

simon-shopify commented:
Updated.

simon-shopify commented:
> I would prefer if we stuck to just using wait_time (and pass that to the block) instead of creating an unnecessary layer of indirection with benchmarks.

oops that was a remnant from the second version… next time when I'm not sure I'll keep old versions around.

csfrancis commented on ext/semian/resource.c:
```diff
@@ -63,7 +63,12 @@ semian_resource_acquire(int argc, VALUE *argv, VALUE self)
+    rb_hash_aset(options, ID2SYM(id_wait_time), INT2NUM(res.wait_time));
+  }
+
+  return rb_ensure(rb_yield, options, cleanup_semian_resource_acquire, self);
```
Why pass a hash here when you could just pass `wait_time`?

simon-shopify commented:
Updated: turned `wait_time` from an option to a positonal parameter.

simon-shopify commented:
> My apologies for not being able to give a very detailed review here. Please just take my comments as suggestions and not requirements.

I guess that means :shipit: ?
